Multi version lsm docs (#7973)
inmanta/inmanta-core: docs/lsm/multi_version/multi_version.rst

**************
Multi-version
**************

> The goal of multi-version lsm is to have versioning as a way to introduce changes to service entities while
> supporting the older behaviour and remaining backwards compatible.
> Each service entity will now have versions, which can introduce breaking changes while maintaining support for older
> versions.

Reviewer: ```suggestion
Multi-version lsm allows you to have multiple api versions for the same service. 
```

> Why do we want to use  multi-version LSM and not different services?

Reviewer: ```suggestion
Why use multi-version LSM?
```

====================================================================
A service represents a high level desired state. With a new version we want to introduce some changes to a service
(i.e. add a new option, or change how some minor things are done), but overall, keep the same intent.
We want to be able to let a service evolve over time while maintaining backwards compatibility and be able to upgrade

> from one version to another without drastically changing the intent of the service.

Reviewer: I would rephrase this to a very practical `when to use multi-version lsm`  section. I think you want to use when 1. you want to offer multiple api schema versions to the same service or 2. you want to upgrade a service in a way that is not supported by the automated upgrade mechanism (if possible, you can link to the docs that describe what is supported, if we have such docs)

Backwards compatibility
=======================

In order to keep backwards compatibility with the v1 endpoints we introduced the concept of a default version.
When we use the v1 endpoints to make operations on a service (i.e. get, create, update), we always target the default
version of the service entity.
Every service entity now has a version. The existing service entities and each new one created with the
``lsm::ServiceEntityBinding`` or ``lsm::ServiceEntityBindingV2`` will have one version numbered 0 (which will be the
default version for this service).

Additionally we can provide a default version to ``lsm::ServiceBinding`` to change which version will be accessible
through the v1 endpoints.

Updating a service entity version is not supported for versioned entities. The new workflow is to create a new version
with the required changes. However, updates to a version 0 of a service entity are allowed, if it is the only version

> of that service.

Reviewer: This is too lsm developer oriented. Backwards compatibility is important to us, probably less to our end-user. 

I think you have to start with explaining which api calls can be used to manipulate multi version instances, then explain that the same endpoints exists without a version in the url and these use the default version.

Author: Made a section for the API calls. Let me know if you want me to go more in-depth in that section

> Migrating from an existing unversioned model

Reviewer: I think this should go before the previous section, so the example is immediately after the when to use

Reviewer: ```suggestion
Using multi version lsm
```

============================================

Using the :ref:`model to create InterfaceIPAssignments <intro_example>` as an example, we will see how we can turn this
entity into a versioned one.

We just need to change our existing ``lsm::ServiceEntityBinding`` into a ``lsm::ServiceBinding``. Set the
``default_version`` to 0 and create a ``lsm::ServiceBindingVersion`` with the information that we had on our previous

> binding.

Reviewer: I would make this a bulleted list or numbered list

**************

> Multi-version

Reviewer: ```suggestion
Multi-version LSM
```

(i.e. different lifecycle).

.. literalinclude:: multi_version/multi_version_sources/multiple_versions.cf
    :linenos:
    :language: inmanta
    :lines: 1-86

In this example we add two new versions. Version 1 links to the same entity model but has a different lifecycle while
Version 2 is a different entity altogether.
When unrolling using ``lsm::all`` we can separate into groups of versions and unroll them as we wish.
In this example, version 1 and 2 will be unrolled together into ``InterfaceIPAssignmentV2`` regardless of their original
version, but version 0 will be unrolled separately into ``InterfaceIPAssignment``).

To remove a version we can delete the corresponding ``lsm::ServiceBindingVersion`` from ``lsm::ServiceBinding.versions``
and recompile and export the new model.

NOTE: We cannot remove service entity versions that have active instances.

API endpoints
=======================
The following API endpoints were added in order to manage versioned services:

* `GET lsm/v2/service_catalog` : List all service entity versions of each defined services entity type in the service catalog

> * `POST lsm/v2/service_catalog`: Create a new service entity type in the service catalog and assign a default version to it.

Reviewer: end-users will never use this directly, not relevant here
```suggestion
```

* `GET lsm/v2/service_catalog/<service_entity>/<version>/schema`: Get the json schema for a service entity version.
* `GET lsm/v2/service_catalog/<service_entity>`: Get all versions of the service entity type from the service catalog.
* `GET lsm/v2/service_catalog/<service_entity>/<version>`: Get one service entity version from the service catalog.

> * `PATCH lsm/v2/service_catalog/<service_entity>`: Update an existing service entity type in the service catalog or add a new one.

Reviewer: end-users will never use this directly, not relevant here
```suggestion
```

* `DELETE lsm/v2/service_catalog/<service_entity>/<version>`: Delete an existing service entity version from the service catalog.
* `GET lsm/v2/service_catalog/<service_entity>/<version>/config`: Get the config settings for a service entity version.
* `POST lsm/v2/service_catalog/<service_entity>/<version>/config`: Set the config for a service entity version.
* `PATCH lsm/v1/service_inventory/<service_entity>/<service_id>/update_entity_version`: Migrate a service instance from one service entity version to another.

> 

Reviewer: I miss the ones to interact with the instances.

Author: There are no new endpoints to interact with instances. When applicable we just have an optional `service_entity_version` argument. I'll add this information here as well.

> The older v1 endpoints are still supported.

Reviewer: ```suggestion
The  v1 endpoints are still supported.
```

Updating a service entity version is not supported for versioned entities. The new workflow is to create a new version
with the required changes. However, updates to a version 0 of a service entity are allowed, if it is the only version
of that service.

Migrating instances between service entity version
==================================================

With the introduction of versions to service entities we might want to migrate existing instances to newer versions.
This can only be done with the `/lsm/v1/service_inventory/<service_entity>/<service_id>/update_entity_version` endpoint
or by calling the `lsm_services_update_entity_version` API method on the Inmanta client.

To do this we need to provide all 3 attribute sets that we want the instance to have on the new entity version.
These attribute sets are validated against the schema of the new version.
We also need to provide the target state that we want to set the instance to.

NOTE: This change is impossible to rollback since we override each attribute set. And each attribute set needs to be
compatible with the target entity version.

> 

Reviewer: I would add a code example here as well (short python script e.g.)
